subsys/power/pm_ctrl.c

```diff
@@ -15,65 +15,35 @@
 #include <logging/log.h>
 LOG_MODULE_DECLARE(power);
 
-struct pm_ctrl_info {
-	int pm_state;
-	atomic_t count;
-};
+static atomic_t power_state_disable_count[SYS_POWER_STATE_MAX];
 
-static struct pm_ctrl_info pm_ctrl[] = {
-	{SYS_PM_LOW_POWER_STATE, ATOMIC_INIT(0)},
-	{SYS_PM_DEEP_SLEEP, ATOMIC_INIT(0)},
-};
-
-void sys_pm_ctrl_disable_state(int state)
+void sys_pm_ctrl_disable_state(enum power_states state)
 {
-	if (state == SYS_PM_LOW_POWER_STATE) {
-		__ASSERT(pm_ctrl[0].count < UINT_MAX,
-				"Low Power state count overflowed\n");
-		atomic_inc(&pm_ctrl[0].count);
-	} else if (state == SYS_PM_DEEP_SLEEP) {
-		__ASSERT(pm_ctrl[1].count < UINT_MAX,
-				"Deep Sleep state count overflowed\n");
-		atomic_inc(&pm_ctrl[1].count);
-	} else {
-		LOG_WRN("\nInvalid PM state");
-	}
+	atomic_val_t v;
+
+	__ASSERT(state < SYS_POWER_STATE_MAX, "Invalid state!\n");
+	v = atomic_inc(&power_state_disable_count[state]);
+	__ASSERT(v < UINT_MAX, "Power state disable count overflowed!\n");
+
+	/* Make compiler happy when assertions are disabled. */
+	(void)(v);
 }
 
-void sys_pm_ctrl_enable_state(int state)
+void sys_pm_ctrl_enable_state(enum power_states state)
 {
-	if (state == SYS_PM_LOW_POWER_STATE) {
-		__ASSERT(pm_ctrl[0].count > 0,
-				"Low Power state count underflowed\n");
-		atomic_dec(&pm_ctrl[0].count);
-	} else if (state == SYS_PM_DEEP_SLEEP) {
-		__ASSERT(pm_ctrl[1].count > 0,
-				"Deep Sleep state count underflowed\n");
-		atomic_dec(&pm_ctrl[1].count);
-	} else {
-		LOG_WRN("\nInvalid PM state");
-	}
+	atomic_val_t v;
+
+	__ASSERT(state < SYS_POWER_STATE_MAX, "Invalid state!\n");
```

Comment:
Same here - we are changing the behavior, as far as I understood....

Reply:
If there will be underflow or overflow, this is pure fault of application, which misuses the API. Without assert this will be unnoticed resulting in undesired behaviour.

Reply:
See above.

Reply:
Fixed